docs/DataFetchingGuide.md

```diff
@@ -0,0 +1,415 @@
+  <source src="./img/AutoRefresh.mp4" type="video/mp4" />
+  Your browser does not support the video tag.
+</video>
+
+The local API mirror significantly enhances both the user experience (with a snappy and responsive UI) and the developer experience (by abstracting caching, invalidation, and optimistic updates).
+
+## Mutation Mode
+
+React-admin provides three approaches for handling updates and deletions:
+
+- **Undoable** (default): React-admin updates the UI immediately and displays an undo button. During this time, it doesn't send a request to the server. If the user clicks the undo button, React-admin restores the previous UI state and cancels the server request. If the user doesn't click the undo button, it sends the request to the server after the delay.
+- **Optimistic**: React-admin updates the UI immediately and sends the request to the server simultaneously. If the server request fails, the UI is reverted to its previous state to maintain consistency.
+- **Pessimistic**: React-admin sends the request to the server first. After the server confirms success, the UI is updated. If the request fails, it displays an error message to inform the user.
+
+![Success message with undo](./img/DeleteButton_success.png)
+
+For each mutation hook or component, you can specify the mutation mode:
+
+```jsx
+const DeletePostButton = ({ record }) => {
+    const [deleteOne] = useDelete(
+        'posts',
+        { id: record.id },
+        { mutationMode: 'pessimistic' }
+    );
+    const handleClick = () => deleteOne();
+    return <Button label="Delete" onClick={handleClick} />;
+};
+```
+
+For details, refer to the [Querying the API](./Actions.md#optimistic-rendering-and-undo) chapter.
+
+## Custom Data Provider Methods
+
+Your API backend may expose non-CRUD endpoints, e.g., for calling Remote Procedure Calls (RPC).
+
+For instance, let's say your API exposes an endpoint to ban a user based on its `id`:
+
+```
+POST /api/user/123/ban
+```
+
+The react-admin way to expose these endpoints to the app components is to add a custom method in the `dataProvider`:
+
+```jsx
+import simpleRestDataProvider from 'ra-data-simple-rest';
+
+const baseDataProvider = simpleRestDataProvider('http://path.to.my.api/');
+
+export const dataProvider = {
+    ...baseDataProvider,
+    banUser: (userId) => {
+        return fetch(`/api/user/${userId}/ban`, { method: 'POST' })
+            .then(response => response.json());
+    },
+}
+```
+
+Then you can use react-query's `useMutation` hook to call the `dataProvider.banUser()` method:
+
+```jsx
+import { useDataProvider } from 'react-admin';
+import { useMutation } from '@tanstack/react-query';
+
+const BanUserButton = ({ userId }) => {
+    const dataProvider = useDataProvider();
+    const { mutate, isPending } = useMutation({
+        mutationFn: () => dataProvider.banUser(userId)
+    });
+    return <Button label="Ban" onClick={() => mutate()} disabled={isPending} />;
+};
+```
+
```

Comment:
A TS example would be better here, showing how to declare your own dataProvider interface and pass it to the `useDataProvider` hook